[Sema] Add global actor attribute to type interface

angela-laar opened the pull request:
Add global actor attribute to type interface when there is an explicit attribute on a declaration. Follow up pull requests will include inferred isolation in the type interface as well. This will help avoid adjusting function types later on during type checking which is prone to errors.

angela-laar commented:
@swift-ci please smoke test

angela-laar committed:
[Sema] Add global actor attribute to type interface

angela-laar commented:
@swift-ci  please smoke test

angela-laar committed:
[Sema] Add global actor attribute to type interface

angela-laar commented:
@swift-ci please smoke test

angela-laar commented:
@swift-ci please smoke test

angela-laar commented:
@swift-ci please smoke test

angela-laar committed:
[Sema] Don't diagnose errors when checking isolation attributes from inside InterfaceTypeRequest

angela-laar committed:
[test] Functions isolated to non-resilient global actors nested in inlinable functions should be diagnosed.

angela-laar commented:
@swift-ci please smoke test

angela-laar commented:
@swift-ci please smoke test

angela-laar committed:
[Sema] Downgrade global_actor_access error to warning for methods

angela-laar commented:
@swift-ci please smoke test

xedin commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -2557,8 +2557,35 @@ InterfaceTypeRequest::evaluate(Evaluator &eval, ValueDecl *D) const {
+            }
+            auto isolation = FunctionTypeIsolation::forGlobalActor(
+                mainIsolation->getGlobalActor());
+            infoBuilder = infoBuilder.withIsolation(isolation);
```
If we consider that the `mainIsolation` is the correct one, we should use it directly, `isolation` produced from it would be the same as `mainIsolation` itself because it only ever returns `forGlobalActor` isolation with `MainActor` type.

angela-laar replied:
mainIsolation is `ActorIsolation` type but `ASTBuilder::withIsolation()` expects `FunctionTypeIsolation` type as it's parameter

angela-laar replied:
Oh actually I see now there is a `ASTExtInfo::withGlobalActor(Type globalActor)` function I can pass `mainIsolation->getGlobalActor()` that will create FunctionTypeIsolation for me

xedin replied:
Right, should just use `mainIsolation->getGlobalActor())` directly

angela-laar replied:
updated

xedin commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -2557,8 +2557,35 @@ InterfaceTypeRequest::evaluate(Evaluator &eval, ValueDecl *D) const {
+      auto &ctx = AFD->getASTContext();
+
+      // Check for explicit global actor attribute
+      auto isolationFromAttr = getIsolationFromAttributes(AFD, false);
```
Regarding `false` argument here is the idea that `getIsolationFromAttributes` would still be called from `ActorIsolationRequest` and be diagnosed there?  If so, what is the plan to address that in the future because we'd like to remove `getIsolationFromAttributes` from `ActorIsolationRequest` right?

hborla replied:
> we'd like to remove getIsolationFromAttributes from ActorIsolationRequest right?

I thought the eventual goal was to directly call `ActorIsolationRequest` from `InterfaceTypeRequest` after we move the diagnostics that require the interface type out from `ActorIsolationRequest`.

xedin replied:
I think the question is still valid though - if we want to call ActorIsolationRequest from InterfaceTypeRequest without diagnostics diagnostics would have to go somewhere.

angela-laar replied:
>  but there is something related to getOverridenDeclarations that makes it recursive when called from interface type?
Yes this is the second reason we can't call `ActorIsolationRequest` directly from `InterfaceTypeRequest`

xedin replied:
Just to be clear I don't think this is a blocker, just trying to understand what the plan is for that because if we don't have test-cases for some situations we'd just loose diagnostics without noticing...

xedin commented on test/attr/attr_inlinable_global_actor.swift:
```diff
@@ -43,10 +44,10 @@ public struct PublicGA {
   //        inlinable functions should be diagnosed.
-  @PrivateGA func inlineFuncPrivateGA() {}
-  @InternalGA func inlineFuncInternalGA() {}
+  @PrivateGA func inlineFuncPrivateGA() {} // expected-error {{struct 'PrivateGA' is private and cannot be referenced from an '@inlinable' function}}
```
Do we want to downgrade this to a warning as well to be safe?

angela-laar committed:
[AST] Set isolation using global actor

If global actor type is availalble, use it to set the isolation in
ASTExtInfo instead of FunctionTypeIsolation
